### PR TITLE
add MaskedTextProviderTests (#7758)

### PR DESCRIPTION
**Pull request by devtwogenies**
Add tests for  MaskedTextProvider related to #7758

### CODE REVIEW

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 1]
```diff
@@ -0,0 +1,640 @@
+﻿using System.Globalization;
+using Xunit;
+using System.Linq;
+using Extensions;
+
+namespace System.ComponentModel.Tests
+{
+    public class MaskedTextProviderTests
+    {
+
```
Comment (maryamariyan) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: nit: extra line

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 2]
```diff
@@ -0,0 +1,640 @@
+﻿using System.Globalization;
```
Comment (maryamariyan) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Add license header to new files:

```
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
```

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+using Xunit;
+using System.Linq;
+using Extensions;
```
Comment (maryamariyan) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: nit: sort usings

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+
+        [Fact]
+        public void VerifyString_Input_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestComplexDigitMask, new CultureInfo("en-US"));
+
+            Assert.True(maskedTextProvider.VerifyString("10000000", out var testPosition, out var resultHint));
+            Assert.Equal(10, testPosition);
+            Assert.Equal(MaskedTextResultHint.Success, resultHint);
+
```
Comment (maryamariyan) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: nit: extra line

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("1", 0));
+            Assert.Equal(1, maskedTextProvider.AssignedEditPositionCount);
+
+            Assert.True(maskedTextProvider.Remove(out var testPosition, out var resultHint));
+            Assert.Equal(MaskedTextResultHint.Success, resultHint);
+            Assert.Equal(0, testPosition);
+        }
+
```
Comment (maryamariyan) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: nit: extra line

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.Equal(MaskedTextResultHint.Success, resultHint);
+
+
+            Assert.False(maskedTextProvider.VerifyString("A", out testPosition, out  resultHint));
+            Assert.Equal(1, testPosition);
+            Assert.Equal(MaskedTextResultHint.DigitExpected, resultHint);
+        }
+
+        [Fact]
+        public void CommonMaskTests()
```
Comment (maryamariyan) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Please break down this test to multiple.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs (hunk 2 above)
Comment (safern) [reply src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Please do it for all the files 😄

File: src/System.ComponentModel.TypeConverter/tests/Extensions/EnumeratorExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+
+namespace Extensions
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/Extensions/EnumeratorExtensions.cs]: Namespace should be `System.ComponentModel.TypeConverter.Tests`

**Comment by maryamariyan** [comment]
thanks @devtwogenies for your contribution. 

As announced in github/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is to apply a patch.

Otherwise if you're planning on updating the PR prior to the November 13th deadline, then patch stage could be skipped.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.Equal(1033, maskedTextProvider.Culture.LCID);
+            Assert.True(maskedTextProvider.AsciiOnly);
+        }
+
+        [Fact]
+        public void Ctor_Mask_PasswordChar_AllowPromptAsInput()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask, '&', false);
+            Assert.Same(TestSimpleDigitMask, maskedTextProvider.Mask);
+            Assert.Equal('&',maskedTextProvider.PasswordChar);
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit:
```suggestion
            Assert.Equal('&', maskedTextProvider.PasswordChar);
```

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 9]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.Same(TestSimpleDigitMask, maskedTextProvider.Mask);
+            Assert.True(maskedTextProvider.AsciiOnly);
+        }
+
+        [Fact]
+        public void Ctor_Mask_Culture()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask, new CultureInfo("en-US"));
+            Assert.Same(TestSimpleDigitMask, maskedTextProvider.Mask);
+            Assert.Equal(1033, maskedTextProvider.Culture.LCID);
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Can we make this `1033` value a constant as well?

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+        }
+
+
+        [Fact]
+        public void Ctor_Mask_Culture_PasswordChar_AllowPromptAsInput()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask, new CultureInfo("en-US"), '&', false);
+            Assert.Same(TestSimpleDigitMask, maskedTextProvider.Mask);
+            Assert.Equal(1033, maskedTextProvider.Culture.LCID);
+            Assert.Equal('&',maskedTextProvider.PasswordChar);
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit:
```suggestion
            Assert.Equal('&', maskedTextProvider.PasswordChar);
```

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+        public void Ctor_Mask_Culture_AllowPromptAsInput_PromptChar_PasswordChar_RestrictToAscii()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask,
+                new CultureInfo("en-US"), false,'-','&', false);
+
+            Assert.Same(TestSimpleDigitMask, maskedTextProvider.Mask);
+            Assert.Equal(1033, maskedTextProvider.Culture.LCID);
+            Assert.False(maskedTextProvider.AllowPromptAsInput);
+            Assert.Equal('-', maskedTextProvider.PromptChar);
+            Assert.Equal('&',maskedTextProvider.PasswordChar);
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit:
```suggestion
            Assert.Equal('&', maskedTextProvider.PasswordChar);
```

Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Can we make `&` and `-` constants?

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 12]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.Equal(1033, maskedTextProvider.Culture.LCID);
+            Assert.Equal('&',maskedTextProvider.PasswordChar);
+            Assert.False(maskedTextProvider.AllowPromptAsInput);
+        }
+
+        [Fact]
+        public void Ctor_Mask_Culture_AllowPromptAsInput_PromptChar_PasswordChar_RestrictToAscii()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask,
+                new CultureInfo("en-US"), false,'-','&', false);
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit:
```suggestion
                new CultureInfo("en-US"), false, '-', '&', false);
```

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 13]
```diff
@@ -0,0 +1,640 @@
...
+
+        [Fact]
+        public void FindAssignedEditPositionInRange()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+            Assert.Equal(-1, maskedTextProvider.FindAssignedEditPositionInRange(0, 1, true));
+
+            maskedTextProvider.Add("12");
+            Assert.Equal(0, maskedTextProvider.FindAssignedEditPositionInRange(0, 1, true));
+            Assert.Equal(2, maskedTextProvider.FindAssignedEditPositionInRange(0, 5,false));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit:
```suggestion
            Assert.Equal(2, maskedTextProvider.FindAssignedEditPositionInRange(0, 5, false));
```

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 14]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.False(MaskedTextProvider.GetOperationResultFromHint(MaskedTextResultHint.PositionOutOfRange));
+        }
+
+        [Fact]
+        public void InsertAt_Input_Position()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+            Assert.False(maskedTextProvider.InsertAt('1', -1));
+            Assert.False(maskedTextProvider.InsertAt('1', 6));
+
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit: remove this extra line.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 15]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.True(maskedTextProvider.InsertAt('1', 0));
+        }
+
+        [Fact]
+        public void InsertAt_Input_Position_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt(string.Empty, 0, out var testPosition,
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: nit: don't use var here.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 16]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.True(maskedTextProvider.InsertAt("1", 0));
+        }
+
+
+        [Fact]
+        public void InsertAt_InputString_Position_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt(string.Empty, 0, out var testPosition,
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Ditto.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 17]
```diff
@@ -0,0 +1,640 @@
...
+
+        [Fact]
+        public void Remove_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("1", 0));
+            Assert.Equal(1, maskedTextProvider.AssignedEditPositionCount);
+
+            Assert.True(maskedTextProvider.Remove(out var testPosition, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit: don't use var here.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 18]
```diff
@@ -0,0 +1,640 @@
...
+
+        [Fact]
+        public void RemoveAt_StartPosition_EndPosition_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("12", 0));
+            Assert.Equal(2, maskedTextProvider.AssignedEditPositionCount);
+
+            Assert.False(maskedTextProvider.RemoveAt(0,6, out var testPosition, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit: 
```suggestion
            Assert.False(maskedTextProvider.RemoveAt(0, 6, out int testPosition, out int resultHint));
```

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 19]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.Equal(6, testPosition);
+            Assert.Equal(MaskedTextResultHint.PositionOutOfRange, resultHint);
+
+            Assert.False(maskedTextProvider.RemoveAt(-1,4, out testPosition, out resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit:
```suggestion
            Assert.False(maskedTextProvider.RemoveAt(-1, 4, out testPosition, out resultHint));
```

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 20]
```diff
@@ -0,0 +1,640 @@
...
+        }
+
+        [Fact]
+        public void Replace_InputChar_Position()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("12", 0));
+            Assert.True(maskedTextProvider.Replace('2', 0));
+            Assert.Equal("2-2_-_", maskedTextProvider.ToDisplayString());
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Make `"2-2_-_"` a constant.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 21]
```diff
@@ -0,0 +1,640 @@
...
+        }
+
+        [Fact]
+        public void Replace_InputChar_StartPosition_EndPosition_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("12", 0));
+            
+            Assert.True(maskedTextProvider.Replace('2', 0, 5, out var testPosition, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit, don't use var here.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 22]
```diff
@@ -0,0 +1,640 @@
...
+        }
+
+        [Fact]
+        public void Replace_InputString_Position()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("12", 0));
+            Assert.True(maskedTextProvider.Replace("23", 0));
+            Assert.Equal("2-3_-_", maskedTextProvider.ToDisplayString());
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Make `"2-3_-_"` a constant.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 23]
```diff
@@ -0,0 +1,640 @@
...
+        }
+
+        [Fact]
+        public void Replace_InputString_Position_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("12", 0));
+            
+            Assert.True(maskedTextProvider.Replace("23", 0, out var testPosition, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit: don't use var here.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 24]
```diff
@@ -0,0 +1,640 @@
...
+        }
+
+        [Fact]
+        public void Replace_InputString_StartPosition_EndPosition_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("12", 0));
+
+            Assert.False(maskedTextProvider.Replace("2", 0, 6, out var testPosition, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit: don't use var here.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 25]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.True(maskedTextProvider.Set("1234"));
+            Assert.Equal("1-23-4", maskedTextProvider.ToDisplayString());
+        }
+
+        [Fact]
+        public void Set_InputString_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.Set("1234", out var testPosition, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Ditto.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 26]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.Equal("1-2_-_", maskedTextProvider.ToString(true, true));
+            Assert.Equal("2_-", maskedTextProvider.ToString(true, true, 2, 3));
+            Assert.Equal("*_-", maskedTextProvider.ToString(false,true, true, 2, 3));
+        }
+
+        [Fact]
+        public void VerifyChar()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+            Assert.True(maskedTextProvider.VerifyChar('1', 0, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Nit: don't use var here.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 27]
```diff
@@ -0,0 +1,640 @@
...
+            Assert.True(maskedTextProvider.VerifyString("10000000"));
+            Assert.False(maskedTextProvider.VerifyString("A"));
+        }
+
+        [Fact]
+        public void VerifyString_Input_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestComplexDigitMask, new CultureInfo("en-US"));
+
+            Assert.True(maskedTextProvider.VerifyString("10000000", out var testPosition, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Ditto.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 28]
```diff
@@ -0,0 +1,640 @@
...
+            maskedTextProvider.Clear();
+            Assert.Equal(0, maskedTextProvider.AssignedEditPositionCount);
+        }
+
+        
+        [Fact]
+        public void Clear_MaskedTextResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+            maskedTextProvider.Clear(out var resultHint);
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Don't use var here.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs [hunk 29]
```diff
@@ -0,0 +1,640 @@
...
+        }
+
+        [Fact]
+        public void Replace_InputChar_Position_TestPosition_ResultHint()
+        {
+            var maskedTextProvider = new MaskedTextProvider(TestSimpleDigitMask);
+
+            Assert.True(maskedTextProvider.InsertAt("12", 0));
+            
+            Assert.True(maskedTextProvider.Replace('2', 0, out var testPosition, out var resultHint));
```
Comment (safern) [inline src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: Don't use var here.

File: src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs (hunk 5 above)
Comment (devtwogenies) [reply src/System.ComponentModel.TypeConverter/tests/MaskedTextProviderTests.cs]: I fixed everything I found. Would you mind telling me what kind of tool you use то check extra lines?

**Review by safern — APPROVED** [review]

**Pull request merged**